src/main/kotlin/no/nav/familie/ef/sak/api/beregning/BeregningService.kt

```diff
@@ -1,17 +1,133 @@
 import org.springframework.stereotype.Service
 import java.math.BigDecimal
 import java.math.RoundingMode
+import java.time.LocalDate
 
 @Service
 class BeregningService {
 
-    fun beregnFullYtelse(beregningRequest: BeregningRequest): List<Beløpsperiode> {
-        return finnGrunnbeløpsPerioder(beregningRequest.stønadFom, beregningRequest.stønadTom).map {
-            Beløpsperiode(it.fraOgMedDato,
-                          it.tilDato,
-                          it.beløp.multiply(BigDecimal(2.25)).divide(BigDecimal(12)).setScale(0, RoundingMode.HALF_UP))
+    fun beregnYtelse(vedtaksperioder: List<Periode>, inntektsperioder: List<Inntektsperiode>): List<Beløpsperiode> {
+
+        validerInnteksperioder(inntektsperioder, vedtaksperioder)
+        validerVedtaksperioder(vedtaksperioder)
+
+
+        val beløpForInnteksperioder = inntektsperioder.flatMap { beregnBeløpForInntekt(it) }
+
+        return vedtaksperioder.flatMap {
+            finnStartDatoOgSluttDatoForBeløpsperiode(beløpForInnteksperioder, it.fradato, it.tildato)
+        }
+    }
+
+
+    private fun beregnBeløpForInntekt(inntektsperiode: Inntektsperiode): List<Beløpsperiode> {
+        val (startDato, sluttDato, inntekt, samordningsfradrag) = inntektsperiode
+        return finnGrunnbeløpsPerioder(startDato, sluttDato).map {
+            val avkortningPerMåned = beregnAvkortning(it.beløp, inntekt).divide(BigDecimal(12))
+                    .setScale(0, RoundingMode.HALF_DOWN)
+
+            val fullOvergangsStønadPerMåned =
+                    it.beløp.multiply(BigDecimal(2.25)).divide(BigDecimal(12)).setScale(0, RoundingMode.HALF_EVEN)
+
+            val beløpFørSamordning =
+                    fullOvergangsStønadPerMåned.subtract(avkortningPerMåned).setScale(0, RoundingMode.HALF_UP)
+
+            val utbetaling = beløpFørSamordning.subtract(samordningsfradrag).setScale(0, RoundingMode.HALF_UP)
+
+
+            val beløpTilUtbetalning = if(utbetaling <= BigDecimal.ZERO) BigDecimal.ZERO else utbetaling
+
+            Beløpsperiode(fraOgMedDato = it.fraOgMedDato,
+                          tilDato = it.tilDato,
+                          beløp = beløpTilUtbetalning,
+                          beregningsgrunnlag = Beregningsgrunnlag(samordningsfradrag = samordningsfradrag,
+                                                                  avkortningPerMåned = avkortningPerMåned,
+                                                                  fullOvergangsStønadPerMåned = fullOvergangsStønadPerMåned,
+                                                                  inntekt = inntekt,
+                                                                  grunnbeløp = it.beløp))
+        }
+    }
+
+    private fun beregnAvkortning(grunnbeløp: BigDecimal, inntekt: BigDecimal): BigDecimal {
+        val inntektOverHalveGrunnbeløp = inntekt.subtract(grunnbeløp.multiply(BigDecimal(0.5)))
+        return if (inntektOverHalveGrunnbeløp > BigDecimal(0))
```

Comment:
Alle disse predefinierte bigdecimal kan nog trekkes ut til egne fields i klassen,
NOE_SOM_FORKLARER_HVA_DET_ER = BigDecimal(0.45)

BigDecimal(0) kan byttes ut med BigDecimal.ZERO, har kanskje ikke så mye å si for vår app, men ellers er det god praksis :)

Reply:
Enig :)

Reply:
Fiksat her [30a9237](https://github.com/navikt/familie-ef-sak/pull/630/commits/30a92379601c378b835a2d8cbd2ce634f66a48c4)